src/main/java/net/sf/jabref/remote/DBProcessor.java

```diff
@@ -0,0 +1,554 @@
+            LOGGER.error(Localization.lang("Corrupt_remote_database_structure."));
+        }
+    }
+
+    /**
+     * Inserts the given bibEntry into remote database.
+     * @param bibEntry {@link BibEntry} to be inserted
+     */
+    public void insertEntry(BibEntry bibEntry) {
+        prepareEntryTableStructure(bibEntry);
+
+        // Check if already exists
+        int remote_id = bibEntry.getRemoteId();
+        if (remote_id != -1) {
+            try (ResultSet resultSet = dbHelper.query(
+                    "SELECT * FROM " + escape(ENTRY) + " WHERE " + escape(ENTRY_REMOTE_ID) + " = " + remote_id)) {
+                if (resultSet.next()) {
+                    return;
+                }
+            } catch (SQLException e) {
+                LOGGER.error("SQL Error: ", e);
+            }
+        }
+
+
+        String query = "INSERT INTO " + escape(ENTRY) + "(";
+        ArrayList<String> fieldNames = new ArrayList<>(bibEntry.getFieldNames());
+
+        for (int i = 0; i < fieldNames.size(); i++) {
+            query = query + escape(fieldNames.get(i).toUpperCase()) + ", ";
+        }
+        query = query + escape(ENTRY_ENTRYTYPE) + ") VALUES(";
+        for (int i = 0; i < fieldNames.size(); i++) {
+            query = query + escapeValue(bibEntry.getField(fieldNames.get(i))) + ", ";
+        }
+        query = query + escapeValue(bibEntry.getType()) + ")";
+
+        try (PreparedStatement preparedStatement = connection.prepareStatement(query,
+                new String[] {ENTRY_REMOTE_ID.toLowerCase(Locale.ENGLISH)})) { // This is the only method to get generated keys which is accepted by MySQL, PostgreSQL and Oracle.
+            preparedStatement.executeUpdate();
+            try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
+                if (generatedKeys.next()) {
+                    bibEntry.setRemoteId(generatedKeys.getInt(1)); // set generated ID locally
+                }
+                preparedStatement.close();
+                generatedKeys.close();
+            }
+        } catch (SQLException e) {
+            LOGGER.error("SQL Error: ", e);
+        }
+
+        LOGGER.info("SQL INSERT: " + query);
+    }
+
+    /**
+     * Updates the hole bibEntry remotely.
+     *
+     * @param bibEntry {@link BibEntry} affected by changes
+     */
+    public void updateEntry(BibEntry bibEntry) {
+        prepareEntryTableStructure(bibEntry);
+
+        String query = "UPDATE " + escape(ENTRY) + " SET " + escape(ENTRY_ENTRYTYPE) + " = "
+                + escapeValue(bibEntry.getType());
+
+        List<String> fields = new ArrayList<>();
+        fields.addAll(bibEntry.getFieldNames());
+
+        List<String> emptyFields = new ArrayList<>();
+        emptyFields.addAll(getRemoteEntry(bibEntry.getRemoteId()).getFieldNames());
+        emptyFields.removeAll(fields); // emptyFields now contains only fields which should be null.
```

Comment:
Now I converted it to `Set<String>`, that made it quit shorter. Unfortunately I did not found a possibility to acces external variable out from a lambda expression. Do you know, how to do it?